src/object/cli_obj.c

```diff
@@ -4758,38 +4759,37 @@ obj_comp_cb(tse_task_t *task, void *data)
 
 		if (task->dt_result == -DER_CSUM || task->dt_result == -DER_TX_UNCERTAIN ||
 		    task->dt_result == -DER_NVME_IO) {
-			if (!obj_auxi->spec_shard && !obj_auxi->spec_group &&
-			    !obj_auxi->no_retry && !obj_auxi->ec_wait_recov) {
+			if (!obj_auxi->spec_shard && !obj_auxi->spec_group && !obj_auxi->no_retry &&
+			    !obj_auxi->ec_wait_recov) {
 				/* Retry fetch on alternative shard */
-				if (obj_auxi->opc == DAOS_OBJ_RPC_FETCH) {
-					if (task->dt_result == -DER_CSUM)
+				if ((obj_auxi->opc == DAOS_OBJ_RPC_FETCH ||
+				     obj_auxi->opc == DAOS_OBJ_RPC_UPDATE) &&
+				    task->dt_result == -DER_CSUM) {
+					struct shard_rw_args *rw_arg = &obj_auxi->rw_args;
+
+					/* Retry a few times on checksum error; something must
+					 * go terribly wrong if checksum happened for 10 times in a
+					 * row therefore we stop trying anyways even though there
+					 * may exist more replicas. */
+					if (rw_arg->csum_retry_cnt < MAX_CSUM_RETRY) {
```

Comment:
for fetch, it should be

if (rw_arg->csum_retry_cnt < max(MAX_CSUM_RETRY, obj_get_replicas(obj)) ?

Reply:
ah, did not notice your comments.